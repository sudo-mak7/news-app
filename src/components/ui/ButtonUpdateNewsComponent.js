import React from 'react'
import { Button, Icon } from 'semantic-ui-react'
import { useDispatch } from 'react-redux'
import { newsGetter } from '../../utils/newsGetter'

const ButtonUpdateNewsComponent = ({ setError }) => {
  const dispatch = useDispatch()

  const errorMessage = 'Error loading news :('

  const updateNews = () => {
    newsGetter(dispatch, setError, errorMessage)
  }

  return (
    <Button
      icon
      size='big'
      color='blue'
      style={{ marginLeft: '1em', zIndex: '999', position: 'fixed' }}
      onClick={ () => updateNews() }
    >
      <Icon name='sync alternate'></Icon>
    </Button>
  )
}

export default ButtonUpdateNewsComponent