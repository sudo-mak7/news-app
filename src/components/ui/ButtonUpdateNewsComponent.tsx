import * as React from 'react'
import { Button, Icon } from 'semantic-ui-react'
import { useAppDispatch } from '@redux/reduxHooks'
import { clearNewsIds, fetchNewsIds } from '@redux/news/newsIdsSlice'
import { clearNews } from '@redux/news/newsSlice'
import { clearPaginationState } from '@redux/pagination/paginationSlice'

const ButtonUpdateNewsComponent = (): JSX.Element => {
  const dispatch = useAppDispatch()

  const updateNews = (): void => {
    dispatch(clearPaginationState())
    dispatch(clearNews())
    dispatch(clearNewsIds())
    dispatch(fetchNewsIds())
  }

  return (
    <Button
      icon
      size='big'
      color='blue'
      style={{
        zIndex: '999',
        position: 'fixed',
        top: '0.25em',
        left: '0.25em'
      }}
      onClick={ (): void => updateNews() }
    >
      <Icon name='sync alternate'></Icon>
    </Button>
  )
}

export default ButtonUpdateNewsComponent