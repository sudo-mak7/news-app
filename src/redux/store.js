import { legacy_createStore as createStore, compose, applyMiddleware } from 'redux'
import createSagaMiddleware from 'redux-saga'
import rootReducer from './reducers/rootReducer'
import { newsWatcher } from './sagas/rootSaga'

const sagaMiddleware = createSagaMiddleware()

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
    : compose

const configureStore = (preloadedState) => createStore(
  rootReducer,
  preloadedState,
  composeEnhancers(applyMiddleware(sagaMiddleware))
)

const store = configureStore({})

sagaMiddleware.run(newsWatcher)

export default store
