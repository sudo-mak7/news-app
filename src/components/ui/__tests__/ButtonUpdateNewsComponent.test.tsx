import * as React from 'react'
import * as reduxHooks from '@redux/reduxHooks'
import { render, screen, fireEvent } from '@testing-library/react'
import { store } from '@redux/store'
import { Provider } from 'react-redux'
import ButtonUpdateNewsComponent from '@components/ui/ButtonUpdateNewsComponent'

const buttonUpdateNewsComponent = (): JSX.Element =>
  <Provider store={ store }>
    <ButtonUpdateNewsComponent />
  </Provider>

const mockedDispatch = jest.spyOn(reduxHooks, 'useAppDispatch')

describe('ButtonUpdateNewsComponent test', () => {
  test('renders ButtonUpdateNewsComponent', () => {
    render(buttonUpdateNewsComponent())
    expect(buttonUpdateNewsComponent()).toMatchSnapshot()

    const buttonInButtonUpdateNewsComponent = screen.getByRole('button')
    expect(buttonInButtonUpdateNewsComponent).toHaveClass('ui blue big icon button')
  })

  test('dispatch should called 4 times by click', () => {
    const dispatch = jest.fn()
    mockedDispatch.mockReturnValue(dispatch)

    render(buttonUpdateNewsComponent())
    const buttonInButtonUpdateNewsComponent = screen.getByRole('button')

    fireEvent.click(buttonInButtonUpdateNewsComponent)
    expect(dispatch).toHaveBeenCalledTimes(4)
  })

  afterAll(() => {
    jest.clearAllMocks()
  })
})