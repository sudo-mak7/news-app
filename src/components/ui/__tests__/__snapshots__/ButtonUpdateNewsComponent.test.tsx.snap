// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonUpdateNewsComponent test renders ButtonUpdateNewsComponent 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ButtonUpdateNewsComponent />
</Provider>
`;
